require(ggplot2)
require(plyr)
require(dplyr)
require(scales)


###################
#Theme for ggplots#
###################
my_gg_theme = theme(panel.background = element_rect(fill='white', colour='black'),
    panel.grid.major=element_blank(),
    panel.grid.minor= element_blank(),
    text=element_text(family="sans"),
    axis.text=element_text(size=15, color="black"),
    axis.ticks = element_line(color = "black"),
    plot.title = element_text(hjust=0, size=20),
    axis.title = element_text(size=17),
    legend.title = element_blank(),
    legend.text = element_text(size = 19),
    strip.text = element_text(size = 15),
    axis.title.x = element_text(margin = margin(t= 10)),
    axis.title.y = element_text(margin = margin(r=10))	)
####################################
#color blind friendly color palette#
####################################
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")



#################
#Load blob tables
#################

#Read blob data (output from combine_taxonomy_coverage_len_GC.pl
blob_dat_1 = read.table("contigs_tax_cov_len_gc.txt")

#Genome wide coverage is useful when comparing assembly quality across multiple assemblies (e.g. different species or different assemblies for the same species)
#The genome-wide coverage value is generated by coverage_per_contig_samtools.pl (grep "genome" coverage_by_sequence.txt)
blob_dat_coverage = 100

#Add this value to the data.frame
blob_dat_1 = data.frame(blob_dat_1, spp = "spp1", Avg_cov = blob_dat_coverage)
#repeat for multiple assemblies


#if plotting data from multiple genomes, first make a data.frame containing all genomes
all_blobs = full_join(blob_dat_1, blob_dat_2) %>%
    full_join(., blob_dat_3) %>%
    full_join(., blob_dat_4) #etc.


#############
### PLOTS ###
#############


###################################
# CLASSIC BLOB PLOT SINGLE GENOME #
###################################


#Make the plot
#The arrange command plots Bacteria first (because they end up last in the dataframe) and are masked if plotting a eukaryote for example
#The arrange command might be changed depending on the dominant organism (e.g. if the sequenced genome is bacterial.
#Alpha value is set in reverse order to Length so that larger points are more translucent

blob_dat_plot = ggplot(arrange(blob_dat_1, desc(Domain)), aes(x = GC, y = Coverage, color = Domain, size = Length, alpha = Length) ) +
geom_point() +
scale_y_log10(limits = c(0.1, 100000), #can be adjusted where appropriate to accomadate Coverage values
    labels = trans_format('log10', math_format(10^.x)) ) +
scale_color_manual(values = cbPalette) +
scale_alpha_continuous(range = c(0.9, 0.33), #Alpha value is set in reverse order to Length so that larger points are more translucent
    trans = "sqrt",
    breaks = c(10^3,10^4,10^5,10^6),
    labels = trans_format('log10', math_format(10^.x)) ) +
scale_x_continuous(limits = c(0,1),
    breaks = c(0, 0.5, 1),
    labels = c("0", "0.5", "1")) +
scale_size_continuous(range = c(0.75,20),#Range sets small and large point size
trans = "sqrt", #Trans can be changed to have a larger or greater variation between point size e.g. trans = "log10", this value must be the same here and in scale_alpha_continuous otherwise two legneds will be plotted
    breaks = c(10^3,10^4,10^5,10^6),
    labels = trans_format('log10', math_format(10^.x)) ) +
labs(size = "Length (bp)",
    alpha = "Length (bp)",
    color = "Taxonomy",
    x = "GC content") +
my_gg_theme +
theme(legend.title = element_text(size = 15)) +
annotate(x = c(0,1), y = c(0.05*Avg_cov, 0.05*Avg_cov ), geom = "line", color = "blue" ) #Plot a line at 5% of avg. genome-wide coverage

#plot figure
pdf("blob_plot.pdf", width = 8, height = 6)
blob_dat_plot
dev.off()

#lower resolution figure
png("blob_plot.png", width = 480, height = 360)
blob_dat_plot
dev.off()

#################################
#CLASSIC BLOBS FACETED BY GENOME#
#################################

classic_blobs_facted = arrange(all_blobs, desc(Domain)) %>% #The arrange command plots Bacteria first (because they end up last in the dataframe)
ggplot(., aes(GC, Coverage, color = Domain, size = Length, alpha = Length) ) + #ggplot command
geom_point() +
facet_wrap(~spp, ncol = as.factor(all_blobs$spp) %>% levels() %>%length()/2 %>% round() ) + #nrow=?
scale_y_log10(limits = c(0.1, 100000),
    labels = trans_format('log10', math_format(10^.x)) ) +
scale_color_manual(values = cbPalette) +
scale_alpha_continuous(range = c(0.9, 0.33), #Alpha value is set in reverse order to Length so that larger points are more translucent
    trans = "sqrt",
    breaks = c(10^3,10^4,10^5,10^6),
    labels = trans_format('log10', math_format(10^.x)) ) +
scale_x_continuous(limits = c(0,1),
    breaks = c(0, 0.5, 1),
    labels = c("0", "0.5", "1")) +
scale_size_continuous(range = c(0.5,10),#Range sets small and large point size
    trans = "sqrt", #Trans can be changed to have a larger or smaller variation between point size e.g. trans = "log10", this value must be the same here and in scale_alpha_continuous otherwise two legends will be plotted
    breaks = c(10^3,10^4,10^5,10^6),
    labels = trans_format('log10', math_format(10^.x)) ) +
labs(size = "Length (bp)",
    alpha = "Length (bp)",
    color = "Taxonomy",
    x = "GC content") +
my_gg_theme +
theme(legend.title = element_text(size = 15))

pdf("faceted_blobs.pdf", width = 16, height = 12)
classic_blobs_facted
dev.off()


#################################
######COVERAGE BY LENGTH PLOT####
#################################
cov_by_length = arrange(blob_dat_1, desc(Domain)) %>%
ggplot(., aes(Coverage/Avg_cov, Length, color = Domain)) + #color sets taxonomic level to color by
geom_point() +
annotate(x = c(0.05, 0.05), #reference line at 5% of avg. genome coverage
    y = c(100, 1000000), #length min max
    geom = "line",
    color = "blue" ) +
annotate(y = c(500, 500), #reference line at 500 bp
    x = c(0, 500), #coverage min max
    geom = "line",
    color = "black") +
my_gg_theme +
scale_y_log10(labels = trans_format('log10', math_format(10^.x)) ) + #log scale axis
scale_x_log10(labels = trans_format('log10', math_format(10^.x)) ) + #log scale axis
xlab(expression(paste(over("Contig coverage", "Avg. genome coverage")))) +
ylab("Contig length (bp)")

pdf("cov_by_len.pdf")
cov_by_length
dev.off()



###########################################
#COVERAGE BY LENGTH PLOT FACETED BY GENOME#
###########################################

cov_by_length_wrapped = arrange(all_blobs, desc(Domain)) %>%
ggplot(., aes(Coverage/Avg_cov, Length, color = Domain)) + #color sets taxonomic level to color by
geom_point(size = 1) +
scale_color_manual(values = cbPalette) +
    annotate(x = c(0.05, 0.05),#reference line at 5% of avg. genome coverage
    y = c(100, 1000000), #length min max
    geom = "line",
    color = "blue" ) +
annotate(y = c(500, 500), #reference line at 500 bp
    x = c(.001, 100), #coverage min max
    geom = "line",
    color = "black") +
facet_wrap(~spp, ncol = as.factor(all_blobs$spp) %>% levels() %>%length()/2 %>% round()) +
my_gg_theme +
scale_y_log10(labels = trans_format('log10', math_format(10^.x)) ) +
scale_x_log10(labels = trans_format('log10', math_format(10^.x)) ) +
xlab(expression(paste(over("Contig coverage", "Avg. genome coverage")))) +
ylab("Contig length (bp)") +
guides(color = guide_legend(override.aes = list(size = 5)))

pdf("faceted_cov_by_len.pdf", width = 12, height = 6)
cov_by_length_wrapped
dev.off()



#######################
#COVERAGE DENSITY PLOT#
#######################
count_cov = ggplot(blob_dat_1, aes(Coverage/Avg_cov, fill = Domain)) + #set fill to color by tax level
geom_bar(stat="bin") +
annotate(x = c(0.05, 0.05), #5% of avg coverage
y = c(0, 50000), #count min max
geom = "line",
color = "blue" ) +
my_gg_theme +
scale_x_log10(labels = trans_format('log10', math_format(10^.x)) ) +
xlab(expression(paste(over("Contig coverage", "Avg. genome coverage")))) +
ylab("Number of contigs") +
theme(axis.title.y = element_text(vjust = 1.75))

pdf("cov_density.pdf")
count_cov
dev.off()


#########################################
#COVERAGE DENSITY PLOT FACETED BY GENOME#
#########################################
count_cov_wrapped = ggplot(all_blobs, aes(Coverage/Avg_cov)) +
geom_bar(stat="bin") +
#annotate(x = c(0.05, 0.05), y = c(0, 10000), geom = "line", color = "blue" ) + #uncomment this line for log-scale y
#scale_y_log10(labels = trans_format('log10', math_format(10^.x)) ) + #uncomment this line for log-scale y
geom_line(stat = "bin", color = "blue", aes(x = 0.05, y = ..count../10)) + #comment this line for log-scale y, this creates a reference line peaking at 5% avg. genome coverage
facet_wrap(~spp,
    ncol = as.factor(all_blobs$spp) %>% levels() %>%length()/2 %>% round(),
    scales = "free_y") +
scale_x_log10(labels = trans_format('log10', math_format(10^.x)) ) +
xlab(expression(paste(over("Contig coverage", "Avg. genome coverage")))) +
ylab("Number of contigs") +
my_gg_theme

pdf("PATH.pdf", width = 12, height = 6)
count_cov_wrapped
dev.off()


